java/arcs/core/host/AbstractArcHost.kt

```diff
@@ -10,43 +10,239 @@
 
-    protected fun registerParticle(particle: ParticleIdentifier) {
-        particles.add(particle)
+    protected fun registerParticle(particle: ParticleIdentifier, constructor: ParticleConstructor) {
+        particles.put(particle, constructor)
     }
 
     protected fun unregisterParticle(particle: ParticleIdentifier) {
         particles.remove(particle)
     }
 
-    override suspend fun registeredParticles(): List<ParticleIdentifier> = particles
+    override suspend fun registeredParticles(): List<ParticleIdentifier> = particles.keys.toList()
+
+    // VisibleForTesting
+    protected fun getArcHostContext(arcId: String) = runningArcs[arcId]
+
+    /** Subclasses may override this to load persistent context state. */
+    protected suspend fun lookupOrCreateArcHostContext(partition: Plan.Partition): ArcHostContext =
+        runningArcs[partition.arcId] ?: ArcHostContext()
+
+    /** Subclasses may override this to store persistent context state. */
+    protected suspend fun updateArcHostContext(arcId: String, context: ArcHostContext) {
+        runningArcs[arcId] = context
+    }
 
     override suspend fun startArc(partition: Plan.Partition) {
-        // TODO: not implemented yet
+        val context = lookupOrCreateArcHostContext(partition)
+
+        // Arc is already currently running, don't restart it
+        if (context.arcState == ArcState.Running) {
+            return
+        }
+
+        // Can't restart a deleted arc
+        if (context.arcState == ArcState.Deleted) {
+            return
+        }
+
+        for (particleSpec in partition.particles) {
+            context.particles[particleSpec] = startParticle(particleSpec, context)
+        }
+
+        performLifecycleForContext(context)
+        maybeRequestResurrection(context)
+
+        context.arcState = ArcState.Running
+        updateArcHostContext(partition.arcId, context)
+    }
+
+    suspend fun startParticle(spec: Plan.Particle, context: ArcHostContext): ParticleContext {
+        val particle = instantiateParticle(ParticleIdentifier.from(spec.location))
+
+        val particleContext = lookupParticleContextOrCreate(
+            context,
+            spec,
+            particle
+        )
+
+        for (handleSpec in spec.handles) {
+            val handle = createHandle(handleSpec.key, handleSpec.value, particle.handles)
+            particleContext.handles[handleSpec.key] = handle
+        }
+
+        return particleContext
+    }
+
+    fun lookupParticleContextOrCreate(
+        context: ArcHostContext,
+        spec: Plan.Particle,
+        particle: Particle
+    ) = context.particles[spec]?.copy(particle = particle) ?: ParticleContext(particle)
+
+    suspend fun performLifecycleForContext(context: ArcHostContext) {
+        for (particleContext in context.particles.values) {
+            performParticleLifecycle(particleContext)
+        }
+    }
+
+    suspend fun performParticleLifecycle(particleContext: ParticleContext) {
```

Comment:
KDoc please. This seems like an important function which should be well-documented for future engineers.